### ISSUE TITLE
Vim9: cannot define function in ":if" block

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#8224]
**Describe the bug**

In Vim9 script, we cannot always define a function inside an `:if` block.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        if true
            def Func()
            enddef
        else
            def Func()
            enddef
        endif
    EOF
    )

`E1073` is raised:

    E1073: Name already defined: Func

**Expected behavior**

No error is raised, because the function is only defined once (in the first clause of the block).

**Environment**

 - Vim version: 8.2 Included patches: 1-2863
 - OS: Ubuntu 20.04.2 LTS
 - Terminal: XTerm(353)

**Additional context**

Regression introduced in 8.2.2864.

---

The error is raised from line 7; that's the footer of the second function definition.
I think it would make more sense if it was raised from line 6, which is the header of the function.

**Comment by lacygoill** [comment]
Among other plugins, this issue breaks the matchparen Vim9 refactoring, where `RemoveHighlight()` is defined inside an `:if` block:
```vim
vim9script
if has('textprop')

  def RemoveHighlight()
    # `:silent!` to suppress E16 in case `line('w$')` is 0
    silent! prop_remove({type: 'matchparen', all: true}, line('w0'), line('w$'))
  enddef

else

  def RemoveHighlight()
    if get(w:, 'matchparen') != 0
      silent! matchdelete(w:matchparen)
      w:matchparen = 0
    endif
  enddef

endif
```

**Comment by lacygoill** [comment]
Probably fixed by 8.2.2865
